### ISSUE TITLE
Please state all dependencies in the readme.

### ISSUE THREAD

**Issue by magnetophon** [ArturKovacs/emulsion#154]
Thanks for making emulsion, looks great!

Unfortunately, I have not been able to run it yet.
I use NixOS, which only makes those dependencies available that are explicitly stated in the nix package.

After compiling for 15 minutes, I got a message saying it needed python3, for xcb, so I added it to the package.
After another 15 minutes I got a message about missing lib xcb, so I added it.

Then it compiled and installed:
```
   Compiling image v0.23.9
   Compiling winit v0.22.0
   Compiling arboard v1.0.1
   Compiling glutin v0.24.1
   Compiling glium v0.27.0
   Compiling gelatin v0.5.0 (/build/source/subcrates/gelatin)
   Compiling emulsion v6.0.0 (/build/source)
warning: unused import: `Rad`
 --> src/widgets/picture_widget.rs:7:32
  |
7 | use gelatin::cgmath::{Matrix4, Rad, Vector3};
  |                                ^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: 1 warning emitted

    Finished release [optimized] target(s) in 5m 38s
     Running target/x86_64-unknown-linux-gnu/release/deps/emulsion-d72814f00191effd

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

installing
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/5vr385q9n73qdfsi2gb5w9c89dyl4knk-emulsion-6.0
shrinking /nix/store/5vr385q9n73qdfsi2gb5w9c89dyl4knk-emulsion-6.0/bin/emulsion
strip is /nix/store/hiwz81i1g3fn3p0acjs042a4h5fri6dh-binutils-2.31.1/bin/strip
stripping (with command strip and flags -S) in /nix/store/5vr385q9n73qdfsi2gb5w9c89dyl4knk-emulsion-6.0/bin
patching script interpreter paths in /nix/store/5vr385q9n73qdfsi2gb5w9c89dyl4knk-emulsion-6.0
checking for references to /build/ in /nix/store/5vr385q9n73qdfsi2gb5w9c89dyl4knk-emulsion-6.0...
building '/nix/store/j0gv2jaxd82yyn5n8qykgxz4z9dmm5aj-user-environment.drv'...
created 1717 symlinks in user environment
profile '/nix/var/nix/profiles/per-user/bart/profile' changed while we were busy; restarting
installing 'emulsion-6.0'
```

When I try to run it however, I get:
```
emulsion                         


Panic happened

--
Failed to initialize any backend! Wayland status: NoCompositorListening X11 status: LibraryOpenError(OpenError { kind: Library, detail: "opening library failed (libX11.so.6: cannot open shared object file: No such file or directory); opening library failed (libX11.so: cannot open shared object file: No such file or directory)" })
--

Location /build/emulsion-6.0-vendor.tar.gz/winit/src/platform_impl/linux/mod.rs:569:9

   0: emulsion::handle_panic::handle_panic
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic
   3: gelatin::application::Application::new
   4: emulsion::main
   5: std::sys_common::backtrace::__rust_begin_short_backtrace
   6: main
   7: __libc_start_main
   8: _start

===================================================================================================
```

So it seems I need to give it libx11.
Anything else I'm missing?

Would it be possible to either check the availability of everything that is needed before starting the compile, or, if that is too much work, list everything needed in the readme, so the next packager has a more easy time?

**Comment by magnetophon** [comment]
I added libX11 to the dependencies, but still got the same error at runtime.
Any ideas?

**Comment by ArturKovacs** [comment]
Hey, thanks for your interest and for opening an issue!

> Anything else I'm missing?

Oh I wish I knew. Would be nice if there would be a way to get `cargo` to list all system-level dependencies but after looking around for a bit I could not find anything of that sort. If anyone could suggest a good method for finding the system-level dependencies I would love to know about it.

The issue you are running into seems to be identical to what's described at #20. @flosse have you ever managed to get emulsion running on NixOS?

**Comment by magnetophon** [comment]
Oh, sorry, that is indeed a practically identical issue.
Let's continue over there!

**Comment by ArturKovacs** [comment]
I actually don't think the two issues are identical even though the phenomenon they describe is identical. This one right now seems to be an ask for help with compilation and execution while #20 is a request to make compilation fail if not all dependecies are installed or to have them installed automatically.

It seems to me that discussing what dependencies are needed would not be appropriate over there.

With that said, installing `libxcursor` and `libxi` might fix this (source: https://github.com/alacritty/alacritty/issues/1742#issuecomment-435616267)

**Comment by magnetophon** [comment]
OK, fair enough.
Feel free to re-open if you want to!

**Comment by magnetophon** [comment]
I added those two libs to the package,  but I'm still getting the same error: ``libX11.so.6: cannot open shared object file: No such file or directory``.

**Comment by ArturKovacs** [comment]
You could try getting a .deb (or one that matches your os but his might not matter) download of Alacrity, unpack the file using your preferred archive manager and check the dependencies there. I tried doing this myself but my archive manager was acting up and I need to go to sleep now. Best of luck

**Comment by ArturKovacs** [comment]
@magnetophon let me know if you made progress with this. Once the dependencies are found I'd like to add them to the readme. I guess I'm going to have to get my hands on a NixOS VM. Although I'm afraid that emulsion on a virtual machine would have graphical issues maybe even panic in cases where it wouldn't on bare metal.

**Comment by magnetophon** [comment]
@ArturKovacs No progress, sorry.
After I looked at Alacritty I decides to put this project in the freezer for now; way [too many dependencies](https://github.com/NixOS/nixpkgs/blob/0f8eb0b8f2e3d440556036d6cae54e9576cb901b/pkgs/applications/misc/alacritty/default.nix), if I really need all of that.

**Comment by ArturKovacs** [comment]
Thanks for sharing this link, this is great to have. We probably don't need all of these, for example the font related ones are almost certainly not needed for now, but most of these are very likely necessary.

I think I'll spend some time over the weekend with trying to make a nix expression for building emulsion.

**Comment by rofrol** [comment]
This is what I got with ldd:

```bash
$ emulsion --version
emulsion 7.2.0
$ ldd ~/.cargo/bin/emulsion 
	linux-vdso.so.1 (0x00007ffc2d3c4000)
	libxcb.so.1 => /lib/x86_64-linux-gnu/libxcb.so.1 (0x00007fd042213000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fd04220d000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fd0421f2000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd042008000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fd041eb9000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fd041eae000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fd041e8a000)
	libXau.so.6 => /lib/x86_64-linux-gnu/libXau.so.6 (0x00007fd041e84000)
	libXdmcp.so.6 => /lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007fd041e7c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fd042a37000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007fd041e62000)
```

**Comment by ArturKovacs** [comment]
Thank you @rofrol using `ldd` was a great idea, which I didn't think about. The problem is that many dynamically linked libraries aren't shown by `ldd`.

As a solution I created a Nix expression under `nix-example` where the dependencies are stored by `rpathLibs`. I updated the readme to explain this for those who want to build emulsion from source. Does this solve the issue for you @magnetophon?

**Comment by magnetophon** [comment]
Thanks!
I'm trying to get it working as a package that I can upstream.
I'm having some issues with the Apple dependencies, but I think I can figure it out.

**Comment by magnetophon** [comment]
It's working now, many thanks!

If I may ask: did you learn nix just for this issue?
I hope you'll get some more use out of it!

**Comment by magnetophon** [comment]
https://github.com/NixOS/nixpkgs/pull/107831

**Comment by ArturKovacs** [comment]
Wow thanks for making a PR for adding it to the official package repo. ❤

> If I may ask: did you learn nix just for this issue?

Haha 😄. The phrase "learn nix" make it sound like I really know my way around Nix which I don't feel like I do; I mostly copy pasted this from alacritty's Nix expression but even with that it did take me quite some time to put the example together.

But yeah this issue, and #20 were the main motivation. I also heard about Nix from an acquaintance and it does seem really cool.

**Comment by magnetophon** [comment]
> Wow thanks for making a PR for adding it to the official package repo. 

Well, *you* did all the work.
I just want binaries built for me from now on!  ;)

> I also heard about Nix from an acquaintance and it does seem really cool.

Well, if the description on the box sounds good to you, I recommend you look into it.
Nix does everything it says on the tin, and then some!


Thanks again!
